tests/cli/net_handler_cli.c

```diff
@@ -0,0 +1,424 @@
+	char nextcmd[MAX_CMD_LEN];
+	
+	//get client to send as
+	while (1) {
+		printf("Send as DAWN or SHEPHERD: ");
+		fgets(nextcmd, MAX_CMD_LEN, stdin);
+		if (strcmp(nextcmd, "dawn\n") == 0) {
+			client = DAWN_CLIENT;
+			break;
+		} else if (strcmp(nextcmd, "shepherd\n") == 0) {
+			client = SHEPHERD_CLIENT;
+			break;
+		} else if (strcmp(nextcmd, "abort\n") == 0) {
+			return;
+		} else {
+			printf("Invalid response to prompt: %s", nextcmd);
+		}
+	}
+	
+	//get pos to send
+	while (1) {
+		printf("Send pos LEFT or RIGHT: ");
+		fgets(nextcmd, MAX_CMD_LEN, stdin);
+		if (strcmp(nextcmd, "left\n") == 0) {
+			pos = LEFT_POS;
+			break;
+		} else if (strcmp(nextcmd, "right\n") == 0) {
+			pos = RIGHT_POS;
+			break;
+		} else if (strcmp(nextcmd, "abort\n") == 0) {
+			return;
+		} else {
+			printf("Invalid response to prompt: %s", nextcmd);
+		}
+	}
+	
+	//send
+	printf("Sending Start Pos message!\n\n");
+	send_start_pos(client, pos);
+}
+
+void prompt_gamepad_state ()
+{
+	uint32_t buttons = 0;
+	float joystick_vals[4];
+	char nextcmd[MAX_CMD_LEN];
+	char **list_of_names;
+	char button_name[16];
+	int button_ix = 0;
+	int set_joysticks = 0;
+	
+	//get which buttons are pressed
+	printf("Specify which of the following buttons are pressed:\n");
+	list_of_names = get_button_names();
+	for (int i = 0; i < NUM_GAMEPAD_BUTTONS; i++) {
+		printf("\t%s\n", list_of_names[i]);
+	}
+	printf("After you have finished setting buttons, type \"done\"\n");
+	printf("If you accidentally set a button, you can unset the button by typing its name again\n");
+	
+	while (1) {
+		printf("Set/unset this button: ");
+		fgets(nextcmd, MAX_CMD_LEN, stdin);
+		
+		//check for these first
+		if (strcmp(nextcmd, "done\n") == 0) {
+			break;
+		} else if (strcmp(nextcmd, "abort\n") == 0) {
+			return;
+		}
+		
+		//get the index of the button that was specified
+		for (button_ix = 0; button_ix < NUM_GAMEPAD_BUTTONS; button_ix++) {
+			//this logic appends a newline to the end of the button name
+			strcpy(button_name, list_of_names[button_ix]);
+			if (strcmp(nextcmd, strcat(button_name, "\n")) == 0) {
+				break;
+			}
+		}
+		if (button_ix == NUM_GAMEPAD_BUTTONS) {
+			printf("Not a button: %s\n", nextcmd);
+			continue;
+		}
+		
+		//otherwise set that bit in buttons
+		if (buttons & (1 << button_ix)) {
+			buttons &= ~(1 << button_ix);
+			printf("\tOK, unset button %s\n", list_of_names[button_ix]);
+		} else {
+			buttons |= (1 << button_ix);
+			printf("\tOK, set button %s\n", list_of_names[button_ix]);
+		}
+	}
+	
+	//get the joystick values to send
+	printf("Specify the values of the four joysticks (values -1.0 <= val <= 1.0):\n");
+	list_of_names = get_joystick_names();
+	for (int i = 0; i < 4; i++) {
+		printf("\t%s\n", list_of_names[i]);
+	}
+	
+	while (set_joysticks != 4) {
+		printf("Set the value for %s: ", list_of_names[set_joysticks]);
+		fgets(nextcmd, MAX_CMD_LEN, stdin);
+		
+		if (strcmp(nextcmd, "abort\n") == 0) {
+			return;
+		}
+		
+		errno = 0;
```

Comment:
Right, so since `0.0` is a valid result for converting a string to a float, the only way you can tell if an error occurred while executing this function is if you first set `errno` to 0, call the function, and then check to see if errno is nonzero and the output is `0.0`. See https://linux.die.net/man/3/strtof , in the "Notes" section...